### PR TITLE
cpu/lm4f120: GPIO support

### PR DESCRIPTION
**Pull request by dkm**
This PR add support for the GPIO API for the lm4f120 cpu

### CODE REVIEW

File: cpu/lm4f120/periph/gpio.c [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+{
+    const uint8_t port_num = _port_num(pin);
+    const uint32_t port_addr = _port_base[port_num];
+
+    return port_addr;
+}
+
+int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+{
+  const uint8_t port_num = _port_num(pin);
```
Comment (thomaseichinger) [inline cpu/lm4f120/periph/gpio.c]: Please spend another 2 spaces for indentation.

File: cpu/lm4f120/periph/gpio.c [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+static void _isr_gpio(uint32_t port_num){
+    const uint32_t port_addr = _port_base[port_num];
+    uint32_t isr = ROM_GPIOPinIntStatus(port_addr, true);
+    uint8_t i;
+
+    ROM_GPIOPinIntClear(port_addr, isr);
+
+    for (i=0; i<8; i++, isr>>=1) {
+        if ((isr & 0x1) == 0)
+            continue;
```
Comment (thomaseichinger) [inline cpu/lm4f120/periph/gpio.c]: Please use brackets also for one line conditional blocks.

File: cpu/lm4f120/periph/gpio.c [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+        if (gpio_config[port_num][i].cb)
+            gpio_config[port_num][i].cb(gpio_config[port_num][i].arg);
```
Comment (thomaseichinger) [inline cpu/lm4f120/periph/gpio.c]: Same as above.

File: cpu/lm4f120/periph/gpio.c [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+        break;
+    }
+
+    ROM_GPIODirModeSet(port_addr, 1<<pin_num, GPIO_DIR_MODE_IN);
+    ROM_GPIOPadConfigSet(port_addr, 1<<pin_num,
+                         GPIO_STRENGTH_2MA, pull_val);
+
+    ROM_IntMasterDisable();
+
+    lm4f120_gpio_int_clear(icr_reg_addr, pin_bit);
```
Comment (thomaseichinger) [inline cpu/lm4f120/periph/gpio.c]: Would you mind replacing this function call by the one liner in the function?

**Comment by dkm** [comment]
@thomaseichinger all done !

**Comment by thomaseichinger** [comment]
Ok great. Let's give Travis a chance.

**Comment by dkm** [comment]
looks good :)

**Comment by dkm** [comment]
refreshed with latest master. I'm waiting for this to be merged before pushing more support for the lm4f120.

**Comment by dkm** [comment]
Any clue as to when this could be merged ? That would be great to be able to get SPI & PWM change along if the target is 03.2016...

**Comment by emmanuelsearch** [comment]
Please rebase!

**Comment by dkm** [comment]
ok, will do this now...

**Comment by dkm** [comment]
any luck this can be merged during the H&A ?

**Comment by emmanuelsearch** [comment]
just tried to toggle LEDs (I'm using the GPIO test https://github.com/RIOT-OS/RIOT/tree/master/tests/periph_gpio).
`toggle` command works when turning on the LEDs (PF1-3) but does not turn them off.
(however it's possible to turn them off with `clear`)

**Comment by dkm** [comment]
ok, will have a look tomorow

**Comment by dkm** [comment]
Ok, fixed it. At least, works correctly here.

File: cpu/lm4f120/include/periph_cpu.h [hunk 5]
```diff
@@ -14,18 +14,31 @@
...
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-/* nothing to do here, yet */
+#define HAVE_GPIO_T
+typedef uint32_t gpio_t;
+
+#define GPIO_PIN(x,y) ((gpio_t)(x<<4 | y))
```
Comment (haukepetersen) [inline cpu/lm4f120/include/periph_cpu.h]: `s/(x<<4 | y)/((x << 4) | y))/

File: cpu/lm4f120/periph/gpio.c [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        pad_config = GPIO_PIN_TYPE_STD_WPD;
+        break;
+    case GPIO_PULLUP:
+        pad_config = GPIO_PIN_TYPE_STD_WPU;
+        break;
+    case GPIO_NOPULL:
+    default:
+        pad_config = GPIO_PIN_TYPE_STD;
+        break;
+    }
```
Comment (haukepetersen) [inline cpu/lm4f120/periph/gpio.c]: Better: Just re-define the values for the `gpio_pp_t` in the `periph_cpu.h` like this:
```c
#define HAVE_GPIO_PP_T
typedef enum {
    GPIO_NOPULL = GPIO_PIN_TYPE_STD,        
    GPIO_PULLUP = GPIO_PIN_TYPE_STD_WPU,        
    GPIO_PULLDOWN = GPIO_PIN_TYPE_STD_WPD
} gpio_pp_t;
```
and you can delete this complete switch/case statement and use the `pullup` variable directly in the next line...

File: cpu/lm4f120/periph/gpio.c [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+                         GPIO_STRENGTH_2MA, pad_config);
+
+    switch(dir){
+    case GPIO_DIR_IN:
+        ROM_GPIODirModeSet(port_addr, pin_bit, GPIO_DIR_MODE_IN);
+        break;
+    case GPIO_DIR_OUT:
+        ROM_GPIODirModeSet(port_addr, pin_bit, GPIO_DIR_MODE_OUT);
+        break;
+    }
```
Comment (haukepetersen) [inline cpu/lm4f120/periph/gpio.c]: same here, just re-define the enum values for `gpio_dir_t` and use `dir` directly in the line below...

File: cpu/lm4f120/periph/gpio.c [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+    case GPIO_BOTH:
+        lm4f_flank_mode = GPIO_BOTH_EDGES;
+        break;
+    case GPIO_RISING:
+        lm4f_flank_mode = GPIO_RISING_EDGE;
+        break;
+    case GPIO_FALLING:
+        lm4f_flank_mode = GPIO_FALLING_EDGE;
+        break;
+    }
```
Comment (haukepetersen) [inline cpu/lm4f120/periph/gpio.c]: and again, just re-define `gpio_flank_t` and use the `flank` var directly

**Comment by dkm** [comment]
Ok, thanks for the feedback. I will take care of this tomorrow !

**Comment by haukepetersen** [comment]
cool, thanks and keep it up :-)

**Comment by thomaseichinger** [comment]
Testing this, output seems to work and also reading works for button SW2 (PF0). As does the external interrupt for this button. For SW1 (PF4) however, neither reading pin level nor interrupts work.

**Comment by dkm** [comment]
hmmm, that's strange as I'm using this as a daily basis since more than a month. I'll test again with a clean build...

**Comment by dkm** [comment]
@thomaseichinger i've just checked-out my branch, flashed the gpio test and it works. But it seems that if you omit to use a pullup, the pin stays low, whereas for SW0, even without pullup, the pins works. Can you confirm that you are using a pullup also ?

**Comment by dkm** [comment]
rebased on top of master + implemented all remarks wrt overriding gpio types.

**Comment by dkm** [comment]
Ok, I guess this one fell back in the low priority list. Ping this PR when you think it's time for it to be merged/reviewed again...

**Comment by emmanuelsearch** [comment]
Travis seems to complain about cpp ?

**Comment by dkm** [comment]
Yes, but I don't see anything in my PR that could cause the standard C++ includes to be missing ?!
I'll try to run the tests locally...

**Comment by gebart** [comment]
The Travis failure makes no sense, it complains on msba2, but this does not touch anything msba2 related.

**Comment by dkm** [comment]
Any way to re-execute the travis build ?

**Comment by dkm** [comment]
Thanks for rescheduling it, travis seems happy now :)

**Comment by gebart** [comment]
I can't test though since I don't have the hardware. I guess someone else will ACK this tonight at the Hack'n'ACK.

**Comment by emmanuelsearch** [comment]
Just tested this PR (building on OS X). 
LEDs (toggling PF1, PF2, PF3) work fine. 
SW2 + PF0 work fine (read, interrupt).
However SW1 + PF4 does not work (not reading anything, no interrupt).
Just to be sure I tested also building with Linux (on Vagrant) but same result SW1 + PF4 do not seem to work.

**Comment by dkm** [comment]
Ok, I've just tried, and it works fine using the gpio test : 

```
init_int 5 4 2 1
```

Not using pull up will cause the button to not work as expected... Can you confirm that you did use a pullup ?

**Comment by emmanuelsearch** [comment]
I confirm SW2 + PF0 works with pull up.

**Comment by emmanuelsearch** [comment]
ACK

**Comment by emmanuelsearch** [comment]
Sorry for the long time it took. Feel free to PR you other stuff now ;)

**Comment by dkm** [comment]
Great, thanks a lot for all the time spent reviewing / testing !

**Pull request merged**